### DISCUSSION TITLE
Weird accelerations at startup

### DISCUSSION THREAD

**Discussion by LuFlo** [JSBSim-Team/jsbsim#642] [Q&A]
I still am fiddling around with the python wrapper. I try to create a very simple scenario. The simulation should start at a resting state, then I would like to perform some inputs programmatically in order to observe the movements, for example starting the engine and throttle up in order to see a forward movement. However, I get some very weird velocities and accelerations at the beginning. I condensed the issue into a short testing script which outputs some basic information about the aircraft every 500 frames:

```python
import jsbsim

PLAY_MAX_FRAMES = 10000


def start_engine(fdm: jsbsim.FGFDMExec):
    fdm.set_property_value('fcs/throttle-cmd-norm', 1.0)
    fdm.set_property_value('propulsion/magneto_cmd', 3)  # = both
    fdm.set_property_value('propulsion/starter_cmd', 1)


def release_brakes(fdm: jsbsim.FGFDMExec):
    fdm.set_property_value('fcs/left-brake-cmd-norm', 0.0)
    fdm.set_property_value('fcs/right-brake-cmd-norm', 0.0)
    fdm.set_property_value('fcs/center-brake-cmd-norm', 0.0)


def reset(fdm: jsbsim.FGFDMExec):
    fdm.set_property_value('velocities/u-fps', 0.0)
    fdm.set_property_value('velocities/v-fps', 0.0)
    fdm.set_property_value('velocities/w-fps', 0.0)
    fdm.set_property_value('attitude/phi-deg', 0.0)
    fdm.set_property_value('attitude/theta-deg', 0.0)
    fdm.set_property_value('attitude/psi-deg', 0.0)
    fdm.set_property_value('ic/vw-north-fps', 0.0)
    fdm.set_property_value('ic/vw-east-fps', 0.0)
    fdm.set_property_value('ic/vw-down-fps', 0.0)

    fdm.set_property_value('position/h-agl-ft', 4.305)  # Took from reset00.xml


def primitive_table_row(*elements) -> str:
    len_row_element = 12
    row = ""
    for el in elements:
        row += max(0, len_row_element - len(el)) * " " + el + "  "
    return row


def reset_accelerations(fdm: jsbsim.FGFDMExec):
    for key, _ in fdm.get_property_catalog('accelerations').items():
        if "gravity" in key:
            continue
        fdm.set_property_value(key, 0.0)


def main():
    fdm = jsbsim.FGFDMExec(None)
    fdm.load_model('c172x')

    fdm.set_property_value("simulation/run_id", 1)
    fdm.set_property_value('fcs/left-brake-cmd-norm', 1)
    fdm.set_property_value('fcs/right-brake-cmd-norm', 1)
    fdm.set_property_value('fcs/center-brake-cmd-norm', 1)

    reset(fdm)
    reset_accelerations(fdm)

    fdm.run_ic()

    frames = -1
    acc_keys = fdm.get_property_catalog('dot-ft_sec2').keys()
    acc_keys_abr = tuple(map(lambda key: key.replace('accelerations', 'a').replace('-ft_sec2', ''), acc_keys))
    table = [primitive_table_row('Lon', 'Lat', 'NED Velocity', 'Airspeed', 'Altitude', 'Frames', 'Time',
                                 *acc_keys_abr)]
    while fdm.run() and ((frames := frames + 1) < PLAY_MAX_FRAMES):
        if frames % 500 == 0 or frames == 0:
            table.append(primitive_table_row(
                f"{fdm['position/long-gc-deg'] : .4f}",
                f"{fdm['position/lat-gc-deg'] : .4f}",
                f"{fdm['velocities/ned-velocity-mag-fps'] : .2f}",
                f"{fdm['velocities/vc-kts'] : .2f}",
                f"{fdm['position/h-agl-ft'] : .3f}",
                f"{frames}",
                f"{fdm.get_sim_time() : .2f}",
                *map(lambda val: f'{val : .2f}', fdm.get_property_catalog('dot-ft_sec2').values())
            ))
        if frames == 5000:
            start_engine(fdm)
            release_brakes(fdm)
    for row in table:
        print(row)
    print(f'Frames executed: {frames}')
    print(f'FPS: {1. / fdm.get_delta_t()}')
    print(f'Sim time: {fdm.get_sim_time()}')


if __name__ == '__main__':
    main()

```

The script outputs this table:
```
         Lon           Lat  NED Velocity      Airspeed      Altitude        Frames          Time        a/udot        a/vdot        a/wdot  
     -0.0000        0.0000         10.82          6.41         0.000             0          0.01          0.00         -1.39       -604.45  
     -0.0000        0.0000         42.10         24.92        55.389           500          4.17        -32.09          6.30         -4.25  
      0.0000       -0.0003         41.34         24.49         7.875          1000          8.34         25.23         14.88         -2.08  
      0.0002       -0.0008         64.84         38.31       181.257          1500         12.51         -4.30         -3.19         -1.45  
      0.0006       -0.0013         38.82         22.99        22.434          2000         16.67        -34.92         -9.42         91.68  
      0.0006       -0.0013          0.56          0.33         4.599          2500         20.84         -1.83          1.91          2.76  
      0.0006       -0.0013          0.01          0.01         4.313          3000         25.01          0.06          0.93          0.05  
      0.0006       -0.0013          0.00          0.00         4.314          3500         29.17          0.00         -0.00          0.00  
      0.0006       -0.0013          0.00          0.00         4.314          4000         33.34          0.00         -0.00         -0.00  
      0.0006       -0.0013          0.00          0.00         4.314          4500         37.51          0.00          0.00          0.00  
      0.0006       -0.0013          0.00          0.00         4.314          5000         41.67          0.00         -0.00          0.00  
      0.0006       -0.0013          8.67          5.14         4.312          5500         45.84          5.33         -0.02          0.08  
      0.0005       -0.0011         30.32         17.96         4.320          6000         50.01          5.03         -0.07          0.10  
      0.0004       -0.0006         50.82         30.11         4.335          6500         54.17          4.80         -0.11          0.16  
      0.0004        0.0000         70.65         41.86         4.359          7000         58.34          4.73         -0.06          0.24  
      0.0006        0.0009         90.26         53.47         4.397          7500         62.51          4.67         -0.02          0.31  
      0.0009        0.0020        109.49         64.86         4.661          8000         66.67          3.77         -0.70          4.19  
      0.0015        0.0032        118.31         70.07        31.222          8500         70.84          3.31         -0.36         -0.52  
      0.0026        0.0034         46.82         27.74         4.172          9000         75.01         -1.58         22.37          2.18  
      0.0024        0.0032          3.16          1.87         4.381          9500         79.18          6.70          3.58          1.74  
```

The engine starts at frame 5000. But way before that the aircraft already experiences huge accelerations. My first guess was, that the aircraft is glitching out of the geod, so I tried to initialize with an altitude of 4.305 (I took the value from a reset file), see line 29. But still no luck, even with higher / lower values. After a while the aircraft seems to rest and it behaves reasonable after the engine starts (I assume a stall at frame 8500).

Does anyone have an idea?

**Comment by seanmcleod** [comment]
In your `reset()` function all the properties you're setting should be `ic/xxx` properties. My guess is that what is happening is that the initial condition for altitude is 0, and so when you start a simulation run the springs of the gear undergo massive compression in order to get the cg of the aircraft to 0ft which then results in the aircraft leaping off the ground etc.

**Reply by LuFlo** [reply]
You saved the day, that was actually the reason! Thanks a lot! I didn't realize that for every `attitude/, velocities/... etc` there is a `ic/...`.

One more thing, do I always need to change the `ic/` properties when I want to "teleport" the aircraft or magically change velocities without inducing accelerations? Or there another way for such changes?

**Reply by bcoconni** [reply]
> One more thing, do I always need to change the `ic/` properties when I want to "teleport" the aircraft or magically change velocities without inducing accelerations? Or there another way for such changes?

One option could be to set the `ic/` properties to new values and then set the property `simulation/reset` to 0 to reset the simulation to initial conditions. This is not guaranteed to succeed however as some of the more complex FCS are not correctly reset. But this might be worth a try !

**Reply by LuFlo** [reply]
Thanks, indeed it is worth a try! Seems easier than reloading the aircraft from scratch.

**Comment by seanmcleod** [comment]
Out of interest, any particular reason you're using:

```python
fdm.set_property_value('ic/vw-north-fps', 0.0)

# Rather than

fdm['ic/vw-north-fps'] = 0.0
```

Also a number that you set are actually read-only.

```c++
  PropertyManager->Tie("velocities/u-fps", this, eU, (PMF)&FGPropagate::GetUVW);
  PropertyManager->Tie("velocities/v-fps", this, eV, (PMF)&FGPropagate::GetUVW);
  PropertyManager->Tie("velocities/w-fps", this, eW, (PMF)&FGPropagate::GetUVW);
```

Note the `(R)` as opposed to `(RW)`.

```python
>>> fdm.query_property_catalog('velocities/u-fps')
['velocities/u-fps (R)']
```

> magically change velocities without inducing accelerations

I doubt JSBSim supports this officially.

**Reply by LuFlo** [reply]
> Out of interest, any particular reason you're using:
> 
> ```python
> fdm.set_property_value('ic/vw-north-fps', 0.0)
> 
> # Rather than
> 
> fdm['ic/vw-north-fps'] = 0.0
> ```

No particular reason, but yeah you are right, the other way looks much cleaner.

> Also a number that you set are actually read-only.
> 
> ```c++
>   PropertyManager->Tie("velocities/u-fps", this, eU, (PMF)&FGPropagate::GetUVW);
>   PropertyManager->Tie("velocities/v-fps", this, eV, (PMF)&FGPropagate::GetUVW);
>   PropertyManager->Tie("velocities/w-fps", this, eW, (PMF)&FGPropagate::GetUVW);
> ```
> 
> Note the `(R)` as opposed to `(RW)`.
> 
> ```python
> >>> fdm.query_property_catalog('velocities/u-fps')
> ['velocities/u-fps (R)']
> ```

Oh yes, now I have replaced the "velocities" with the "ic/" ones, which are RW. When I looked the keys up, I didn't use `query_property_catalog`, I used `get_property_catalog` which strips the R / RW part. But it makes sense, when you are outputting a dictionary, you want to have "clean" keys.

> > magically change velocities without inducing accelerations
> 
> I doubt JSBSim supports this officially.

Well, I guess one use some tricks, like creating a completely new model with new initialization if you want like a scene change or something and let the visual engine handle a seamless transition if necessary. Shouldn't be much of an issue.
